### ISSUE TITLE
Is this a bad error:

### ISSUE THREAD

**Issue by darkoverlordofdata** [plaurent/gnustep-build#16]

**Comment by darkoverlordofdata** [comment]
Sorry, I'm not sure about these error messages, are they fatal or can they be ignored?

```
Compiling and running blocks demo.
Hello, block!
Compiling and running Grand Central Dispatch demo.
NSAutoreleasePool does not support ARC correctly (implements retain)
GSPlaceholderString does not support ARC correctly (implements retain)
GSMutableString does not support ARC correctly (implements retain)
NSConstantString does not support ARC correctly (implements retain)
GSTinyString does not support ARC correctly (implements retain)
The fraction is: 1/3
GSPlaceholderValue does not support ARC correctly (implements retain)
Compiling and running Fraction demo (makefile compilation).
This is gnustep-make 2.8.0. Type 'make print-gnustep-make-help' for help.
Running in gnustep-make version 2 strict mode.
Making all for app FractionDemo...
 Compiling file Fraction.m ...
 Compiling file helloGCD_objc.m ...
 Linking app FractionDemo ...
NSAutoreleasePool does not support ARC correctly (implements retain)
GSPlaceholderString does not support ARC correctly (implements retain)
GSMutableString does not support ARC correctly (implements retain)
NSConstantString does not support ARC correctly (implements retain)
GSTinyString does not support ARC correctly (implements retain)
The fraction is: 1/3
GSPlaceholderValue does not support ARC correctly (implements retain)
Compiling and running Fraction ARC demo (makefile compilation).
This is gnustep-make 2.8.0. Type 'make print-gnustep-make-help' for help.
Running in gnustep-make version 2 strict mode.
Making all for app FractionArcDemo...
 Compiling file Fraction.m ...
 Compiling file helloGCD_ARC_objc.m ...
 Linking app FractionArcDemo ...
NSAutoreleasePool does not support ARC correctly (implements retain)
GSPlaceholderString does not support ARC correctly (implements retain)
GSMutableString does not support ARC correctly (implements retain)
NSConstantString does not support ARC correctly (implements retain)
GSTinyString does not support ARC correctly (implements retain)
The fraction is: 1/3
GSPlaceholderValue does not support ARC correctly (implements retain)
```

**Comment by plaurent** [comment]
I haven’t run into these errors yet. What OS/distribution are you building under to get these errors?

**Comment by plaurent** [comment]
(The error doesn’t seem particularly bad, it seems more like a warning. But it’d be good to dig into what’s going on here.  I’m wondering if somehow libs-base on your system is being compiled with ARC — which libs-base should not be.)

**Comment by darkoverlordofdata** [comment]
clang version 10.0.0-4ubuntu1 
I'm running on Pop!_OS 20.04 LTS, which is based on the corresponding Ubuntu, and I used your script 'ubuntu-20.04-clang-10.0-runtime-2.0' to install gnustep. 

I compiled and installed the gnustep GMines example. It prints out a slew of similar messages, but it appears to run ok. My assumption is that it would not continue to run if this were fatal but I don't really have much experience with objective-c; I'm attempting to learn.

**Comment by plaurent** [comment]
From the discuss-gnustep mailing list we have confirmation those errors are harmless, and sounds like they will soon be hidden in an update to libobjc2. 



On May 20, 2020, at 02:05, David Chisnall wrote:

﻿Hi Andreas,

These warnings are fine to ignore (they should be guarded by debug builds of libobjc2 or an environment variable, but they aren't currently).  You're seeing them now because Richard found and half fixed a bug in our detection of when to use the fast paths, and I finished and merged the patch about a week ago.

The warning is badly worded (my fault) and just means that the class will not use fast paths for ARC.  A few of these could easily be made to go away (NSNull, GSDateSingle, and a few others are singletons and implement -retain defensively to avoid retain calls.  The runtime now provides a better mechanism for doing this, but I haven't yet exposed it.  The ARC implementation has a notion of persistent objects [e.g. NSConstantString] for which it skips all ARC-related calls.  Opting in to this will let NSObject's -retain / -release become no-ops if they're every called).

Some of the others are implementing -retain to copy.  These are strictly wrong.  The spec for -retain requires that it returns self.  Code compiled with ARC may or may not do the right thing here (depending on your compilation flags, the compiler may use the return value of objc_retain, but it may ignore it.  objc_retain returns the argument so that the compiler can avoid spilling the value across calls - it's guaranteed to be in the return register).

I believe the small objects don't need to implement any of these methods.  The runtime should be hitting fast paths for them and not calling memory management methods and if you call NSObject's -retain / -release / -autorelease then these should be delegating to the runtime for any GNUstep build that targets a runtime that supports small objects.

David

On 20/05/2020 08:45, Andreas Fink wrote:
Hello all,
I recompiled the latest version of gnustep yesterday with clang-10 and now I see the following error messages at startup.
NSAutoreleasePool does not support ARC correctly (implements retain)
GSPlaceholderString does not support ARC correctly (implements retain)
GSMutableString does not support ARC correctly (implements retain)
NSConstantString does not support ARC correctly (implements retain)
GSTinyString does not support ARC correctly (implements retain)
GSPlaceholderValue does not support ARC correctly (implements retain)
GSMutableArray does not support ARC correctly (implements retain)
GSPlaceholderArray does not support ARC correctly (implements retain)
GSPlaceholderTimeZone does not support ARC correctly (implements retain)
NSLocalTimeZone does not support ARC correctly (implements retain)
_NSConcreteProcessInfo does not support ARC correctly (implements retain)
GSDateSingle does not support ARC correctly (implements retain)
GSDateFuture does not support ARC correctly (implements retain)
NSSmallInt does not support ARC correctly (implements retain)
NSBundle does not support ARC correctly (implements release)
NSNull does not support ARC correctly (implements retain)
all my code is compiled with ARC which shouldnt stop clang calling code which is non ARC.
why do we see these errors and whats the workaround?
And is there any memory management problem here really or can it be safely ignored.

**Comment by darkoverlordofdata** [comment]
Well, I tried the script for Debian 10. I get those same messages. 

HelloWorld works great, but my game, which works on FreeBSD, gets one set of runtime errors on Ubuntu, and another set of errors on Debian. So I tried porting to ObjFW - all I actually had to do was rename NSString to OFString, etc. And now my code is working on all three systems, so I guess I'll stick with that.

I can see why GNUstep doesn't get used much.
